### ISSUE TITLE
Ex5に｢誤差を気にしなくて良い｣と書いているにも関らず実際は気にする必要がある

### ISSUE THREAD

**Issue by ncaq** [haskell-jp/makeMistakesToLearnHaskell#78]
課題に書かれている｢複利計算の式｣を使えば良いのですが

再帰を使って計算したり,

[taashiさんのツイート: "コード calcCI :: Double -> Double -> Double -> Double calcCI _ 0 a = a calcCI r y a = calcCI r (y - 1) $ a * (1 + r / 100)"](https://twitter.com/taashi_s/status/1061086453382303745)

`^`の代わりに`**`を使うと誤差が発生するようです.

誤差を許容する修正をしようとしたのですが,パット見テキストの同値性を使ってるので即座の修正は難しそうですね.

**Comment by igrep** [comment]
そうですね。。。基本的に初心者をターゲットにして「習ってないことを使わないでね」という方針で作っているので、どこかにそれを明記する形でサポートですかね。。。

**Comment by fumieval** [comment]
数値計算に丸め誤差はつきものなので、この類のテストは閾値を決め、ある程度の誤差を許すのが定石ですね。社内のDiffベースのテストも、一旦Floatにキャストしたものを出力するというアプローチを取っています

**Comment by igrep** [comment]
いっそ `round` 関数を紹介してしまうかな。。。
（それで解決する問題なのか確信が持てませんが）

**Comment by ncaq** [comment]
いっそのこと有理数を使ってもらうのも手では(使えるのか?)

[Data.Ratio](https://www.stackage.org/haddock/lts-12.17/base-4.11.1.0/Data-Ratio.html)

**Comment by igrep** [comment]
なるほど。型注釈で `:: Rational`と書けばいけるかも。

**Comment by ncaq** [comment]
とりあえず現状は入力が小数形式だから使えないですが
金の計算なら誤差が生じない有理数を使うほうがシチュエーションとしてはむしろ正しそう

**Comment by ncaq** [comment]
1. answerとユーザの出力を切り取って最初の10行ぐらいしか見ないようにする
2. テスト関数を単純なテキスト比較ではなく比較関数を受け取るようにして誤差を許容する比較をする
3. 金の計算で誤差が発生する浮動小数点数使うの変なので,いっそのこと課題を有理数を使うものに全面的に書き換える

などの道があると思います

**Comment by igrep** [comment]
と、思ったら `read` 関数でエラーになった  :cry:

**Comment by igrep** [comment]
こちらですが、直し方としては[runHaskellExercise'](https://github.com/haskell-jp/makeMistakesToLearnHaskell/blob/da2c0adda6759629a3d263b9bb82993aaf76ef30/src/Education/MakeMistakesToLearnHaskell/Exercise/Core.hs#L39)や[runHaskellExerciseWithStdin ](https://github.com/haskell-jp/makeMistakesToLearnHaskell/blob/da2c0adda6759629a3d263b9bb82993aaf76ef30/src/Education/MakeMistakesToLearnHaskell/Exercise/Core.hs#L73)に、解答が等しいかどうかを `==` 以外で判定するための関数を引数として追加し、それに `Text.take 5 right == Text.take 5 out` みたいな関数を渡せばできるんじゃないかと踏んでいます。
